### PR TITLE
persist and recover shard states.

### LINKED ISSUE #13: mounts: revisit the way that we provide environmental parameters
Right now we rely on a MountFactory, which can be instantiated with parameters like a JSON-RPC API endpoint, credentials, etc. that are applied to all Mounts created from it. However, it's possible that these parameters will change at runtime, and the current model is too inflexible to accommodate that.

### LINKED ISSUE #24: upgrader: expose the path of the transient
Add a method `TransientPath() string` to the `mount.Upgrader` so we can obtain the local transient path to persist it in shard state. Thus, on a restart, the system should be capable of reusing that file instead of having to fetch again from the remote mount.

### LINKED ISSUE #32: implement shard state persistence and recovery upon restart
The DAG store management layer must keep its internal state synchronised with an embedded kv database, injected by the user through the existing `ds.Datastore` constructor option.

Upon restart, the DAG store must resume its state from the database. On every internal operation, the DAG store must update the shard record in the DB through an upsert.

## Data model

- key => shard key
- value => JSON doc (for debuggability right now; later to be moved to CBOR doc for efficiency)
    - mount URL
    - path to transient
    - last state
    - error msg

## Resumption procedure

Iterate through kv store, and add a shard for every key, populating it with the persisted data.
For active shards, verify that the index exists.
For failed shards, track them as failed and recover the last known error.

Shards are never resumed into Serving state, and their refcount is always 0.

### PR DESCRIPTION
**Pull request by raulk**
Closes #32.
Closes #24.
Closes #13.

This PR fleshes out another important part of the system: shard state persistence and resumption upon restart, and it also rewrites the mount registry/factory abstraction.

## Persisting shard state

On every iteration of the event loop, we persist the current shard state. This is persisted in a `datastore.Datastore` provided by the user. We assume it to be namespaced.

Shards are persisted under their shard key, and the state is serialised as JSON. There is unused code for switching to CBOR in the future, if we find that to be more performant.

We can put on extra smarts to avoid persisting when unnecessary (e.g. when the persisted entity would be identical than the last one written). We can easily do this because `PersistentShard` is a comparable struct, so we could retain the last written `PersistentShard` in the shard state, calculate the new one, and skip the datastore put if they're identical. TODO in https://github.com/filecoin-project/dagstore/issues/44.

We also need to clean up the serialization logic. I don't like `MarshalJSON` and `UnmarshalJSON` doing this much.

## Resuming the DAG store

On start, we restore the state from the Datastore. We iterate and recover all shards, overwriting some interim states:

- ShardStateServing is reset to ShardStateAvailable, as when we start, there is guaranteed to be 0 active consumers.
- ShardStateInitializing is reset to ShardStateNew, and we restart the shard initialization by dispatching tasks to the event loop.

## Mount registry and factory

I revisited the `mount.Registry` and `mount.MountFactory` design. This has been simplified by removing the latter. Instead, we use a `Mount` instance as a template, and we clone it every time we instantiate a new mount of that type.

This allows us to set "environmental" properties that get applied automatically to all mounts of that type, as for example a JSON-RPC endpoint in the Lotus mount:

```go
type LotusMount string {
    Endpoint string
}

var _ mount.Mount = (*LotusMount)(nil) // assume it implements the interface

r := mount.NewRegistry()
r.Register("lotus", &LotusMount{Endpoint: "ws://host:port/rpc/v0"})

m, _ = r.Instantiate(&url.URL{Scheme: "lotus", ...}) // m will have Endpoint set
```

An example of this is used in the tests with the new `mount.FSMount`, which I had to replace `mount.BytesMount` with for testing, since we can't afford to serialize full CAR files into base64 when encoding URLs every time that we persist shards.

Finally, the new registry design allows us to register the same mount type under many different schemes, with different templates. This is important if, for example, you have various `fs.FS` you want to serve from (for testing), or you want to cater to more than one Lotus deployment.

Since we do type matching, you do need to create new types when doing that:

```go
type LotusMount1 struct { *LotusMount }
type LotusMount2 struct { *LotusMount }
type LotusMount3 struct { *LotusMount }

r := mount.NewRegistry()
r.Register("lotus1", &LotusMount1{&LotusMount{Endpoint: "ws://host1:port1/rpc/v0"}})
r.Register("lotus2", &LotusMount2{&LotusMount{Endpoint: "ws://host2:port2/rpc/v0"}})
r.Register("lotus3", &LotusMount3{&LotusMount{Endpoint: "ws://host3:port3/rpc/v0"}})
```

### CODE REVIEW

File: mount/upgrader.go [hunk 1]
```diff
@@ -4,39 +4,41 @@ import (
...
 
-// Upgrade constructs a new Upgrader for the underlying Mount.
-func Upgrade(underlying Mount, initial string) (*Upgrader, error) {
-	ret := &Upgrader{underlying: underlying}
+// Upgrade constructs a new Upgrader for the underlying Mount. If provided, it
+// will reuse the file in path `initial` as the initial transient copy. Whenever
+// a new transient copy has to be created, it will be created under `rootdir`.
+func Upgrade(underlying Mount, rootdir, initial string) (*Upgrader, error) {
+	ret := &Upgrader{underlying: underlying, rootdir: rootdir}
+	if ret.rootdir == "" {
```
Comment (aarshkshah1992) [inline mount/upgrader.go:34]: Given that we ensure that the `TransientDir` exists and is Statable when we create the DAGStore, we should fail here.

**Review by aarshkshah1992 — APPROVED** [review]
Great stuff ! Just some nits.

- Needs some solid tests.
- I'm worried about the fsm Idempotency after resetting the state to `ShardStateNew` but I'll write a test for that.

File: mount/upgrader.go (hunk 1 above)
Comment (raulk) [reply mount/upgrader.go:34]: This is a public component in the dagstore API and can be used independent of the dagstore itself, so we shouldn't couple it with the dagstore's usage.

File: mount/upgrader.go [hunk 2]
```diff
@@ -94,6 +95,23 @@ func (u *Upgrader) Stat(ctx context.Context) (Stat, error) {
...
+}
+
+func (u *Upgrader) Serialize() *url.URL {
+	return u.underlying.Serialize()
+}
+
+func (u *Upgrader) Deserialize(url *url.URL) error {
+	return u.underlying.Deserialize(url)
+}
+
```
[the remark this answers is not in the document]
Comment (raulk) [reply mount/upgrader.go:114]: We don't want to delete the transient on Close, we want it to survive restarts (as it's also persisted in the store). 

note: `mount.Close()` will get called on a graceful shutdown (#34).

File: mount/upgrader.go (hunk 1 above)
Comment (aarshkshah1992) [reply mount/upgrader.go:34]: Oh sorry, didn't realise it's exported. This makes sense.

File: shard.go [hunk 3]
```diff
@@ -40,5 +41,5 @@ type Shard struct {
 	wAcquire  []*waiter
 	wDestroy  *waiter
 
-	refs uint32 // count of DAG accessors currently open
+	refs uint32 // number of DAG accessors currently open
```
[the remark this answers is not in the document]
Comment (raulk) [reply shard.go:44]: Will add.

File: shard_persist.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	Key           string     `json:"k"`
+	URL           string     `json:"u"`
+	State         ShardState `json:"s"`
+	Indexed       bool       `json:"i"`
+	TransientPath string     `json:"t"`
+	Error         string     `json:"e"`
+}
+
+// MarshalJSON returns a serialized representation of the state. It must be
+// called from inside the event loop, as it accesses mutable state.
```
[the remark this answers is not in the document]
Comment (raulk) [reply shard_persist.go:25]: True, well spotted!

File: shard_persist.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	mnt, err := s.d.mounts.Instantiate(u)
+	if err != nil {
+		return fmt.Errorf("failed to instantiate mount from URL: %w", err)
+	}
+	s.mount, err = mount.Upgrade(mnt, s.d.config.TransientsDir, ps.TransientPath)
+	if err != nil {
+		return fmt.Errorf("failed to apply mount upgrader: %w", err)
+	}
+
+	s.indexed = ps.Indexed
```
[the remark this answers is not in the document]
Comment (raulk) [reply shard_persist.go:78]: Yeah, that simplification is worth it. We can use `StatFullIndex` to check if the index exists.

[the remark this answers is not in the document]
Comment (raulk) [reply shard_persist.go:78]: https://github.com/filecoin-project/dagstore/issues/47

**Pull request merged**